File: clients/java/src/test/java/dev/zvn/powdb/ProtocolTest.java
package dev.zvn.powdb;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class ProtocolTest {

    private Message roundtrip(Message in) {
        byte[] bytes = Protocol.encode(in);
        Protocol.Decoded d = Protocol.tryDecode(bytes, bytes.length);
        assertNotNull(d);
        assertEquals(bytes.length, d.consumed());
        return d.message();
    }

    @Test
    void connectWithPassword() {
        Message out = roundtrip(new Message.Connect("default", "secret"));
        Message.Connect c = (Message.Connect) out;
        assertEquals("default", c.dbName());
        assertEquals("secret", c.password());
    }

    @Test
    void connectNoPassword() {
        Message out = roundtrip(new Message.Connect("default", null));
        assertNull(((Message.Connect) out).password());
    }

    @Test
    void connectOk() {
        Message out = roundtrip(new Message.ConnectOk("0.1.2"));
        assertEquals("0.1.2", ((Message.ConnectOk) out).version());
    }

    @Test
    void query() {
        Message out = roundtrip(new Message.Query("User filter .age > 30"));
        assertEquals("User filter .age > 30", ((Message.Query) out).query());
    }

    @Test
    void resultRows() {
        List<List<String>> rows = Arrays.asList(
                Arrays.asList("Alice", "30"),
                Arrays.asList("Bob", "25"));
        Message out = roundtrip(new Message.ResultRows(Arrays.asList("name", "age"), rows));
        Message.ResultRows r = (Message.ResultRows) out;
        assertEquals(Arrays.asList("name", "age"), r.columns());
        assertEquals(rows, r.rows());
    }

    @Test
    void resultRowsEmpty() {
        Message out = roundtrip(new Message.ResultRows(
                Collections.singletonList("x"), Collections.emptyList()));
        assertTrue(((Message.ResultRows) out).rows().isEmpty());
    }

    @Test
    void resultScalar() {
        Message out = roundtrip(new Message.ResultScalar("42"));
        assertEquals("42", ((Message.ResultScalar) out).value());
    }

    @Test
    void resultOk() {
        Message out = roundtrip(new Message.ResultOk(1L << 40));
        assertEquals(1L << 40, ((Message.ResultOk) out).affected());
    }

    @Test
    void errorMessage() {
        Message out = roundtrip(new Message.Error("table not found"));
        assertEquals("table not found", ((Message.Error) out).message());
    }

    @Test
    void disconnect() {
        Message out = roundtrip(new Message.Disconnect());
        assertTrue(out instanceof Message.Disconnect);
    }

    @Test
    void partialFrameReturnsNull() {
        byte[] full = Protocol.encode(new Message.Query("hello"));
        assertNull(Protocol.tryDecode(full, 3));
        assertNull(Protocol.tryDecode(full, full.length - 1));
    }

    @Test
    void utf8Roundtrip() {
        Message out = roundtrip(new Message.ResultScalar("café ☕ 日本語"));
        assertEquals("café ☕ 日本語", ((Message.ResultScalar) out).value());
    }

    @Test
    void unknownTypeThrows() {
        byte[] frame = Protocol.encode(new Message.Disconnect());
        frame[0] = (byte) 0xFF;
        assertThrows(ProtocolException.class, () -> Protocol.tryDecode(frame, frame.length));
    }
}

File: clients/go/protocol_test.go
package powdb

import (
	"reflect"
	"testing"
)

func roundtrip(t *testing.T, in Message) Message {
	t.Helper()
	bytes := Encode(in)
	out, consumed, err := TryDecode(bytes)
	if err != nil {
		t.Fatalf("decode failed: %v", err)
	}
	if out == nil {
		t.Fatalf("decode returned nil message")
	}
	if consumed != len(bytes) {
		t.Fatalf("consumed %d bytes, expected %d", consumed, len(bytes))
	}
	return out
}

func TestConnectRoundtripWithPassword(t *testing.T) {
	pw := "secret"
	out := roundtrip(t, Connect{DBName: "default", Password: &pw})
	c := out.(Connect)
	if c.DBName != "default" || c.Password == nil || *c.Password != "secret" {
		t.Fatalf("unexpected: %+v", c)
	}
}

func TestConnectRoundtripNoPassword(t *testing.T) {
	out := roundtrip(t, Connect{DBName: "default", Password: nil})
	c := out.(Connect)
	if c.Password != nil {
		t.Fatalf("expected nil password, got %v", *c.Password)
	}
}

func TestConnectOk(t *testing.T) {
	out := roundtrip(t, ConnectOk{Version: "0.1.2"})
	if out.(ConnectOk).Version != "0.1.2" {
		t.Fatal("version mismatch")
	}
}

func TestQuery(t *testing.T) {
	out := roundtrip(t, Query{Query: "User filter .age > 30"})
	if out.(Query).Query != "User filter .age > 30" {
		t.Fatal("query mismatch")
	}
}

func TestResultRows(t *testing.T) {
	in := ResultRows{
		Columns: []string{"name", "age"},
		Rows:    [][]string{{"Alice", "30"}, {"Bob", "25"}},
	}
	out := roundtrip(t, in).(ResultRows)
	if !reflect.DeepEqual(out, in) {
		t.Fatalf("mismatch: %+v vs %+v", out, in)
	}
}

func TestResultRowsEmpty(t *testing.T) {
	in := ResultRows{Columns: []string{"x"}, Rows: [][]string{}}
	out := roundtrip(t, in).(ResultRows)
	if len(out.Rows) != 0 {
		t.Fatalf("expected 0 rows, got %d", len(out.Rows))
	}
}

func TestResultScalar(t *testing.T) {
	out := roundtrip(t, ResultScalar{Value: "42"}).(ResultScalar)
	if out.Value != "42" {
		t.Fatal("value mismatch")
	}
}

func TestResultOk(t *testing.T) {
	out := roundtrip(t, ResultOk{Affected: 1 << 40}).(ResultOk)
	if out.Affected != 1<<40 {
		t.Fatalf("affected mismatch: %d", out.Affected)
	}
}

func TestError(t *testing.T) {
	out := roundtrip(t, Error{Message: "table not found"}).(Error)
	if out.Message != "table not found" {
		t.Fatal("message mismatch")
	}
}

func TestDisconnect(t *testing.T) {
	out := roundtrip(t, Disconnect{})
	if _, ok := out.(Disconnect); !ok {
		t.Fatalf("expected Disconnect, got %T", out)
	}
}

func TestPartialFrameReturnsNil(t *testing.T) {
	full := Encode(Query{Query: "hello"})
	for _, truncLen := range []int{0, 3, len(full) - 1} {
		msg, consumed, err := TryDecode(full[:truncLen])
		if err != nil || msg != nil || consumed != 0 {
			t.Fatalf("truncated[%d]: expected (nil, 0, nil), got (%v, %d, %v)", truncLen, msg, consumed, err)
		}
	}
}

func TestUTF8Roundtrip(t *testing.T) {
	out := roundtrip(t, ResultScalar{Value: "café ☕ 日本語"}).(ResultScalar)
	if out.Value != "café ☕ 日本語" {
		t.Fatalf("utf8 mismatch: %q", out.Value)
	}
}
